bert_pretrain/README.md

```diff
@@ -7,34 +7,26 @@
 
 # Mosaic BERT
 
-This folder contains starter code for training BERT with Composer. Our goal was to apply a combination of tricks to seriously improve training time, and to package it a way that's easy for you to use on your own problems. We apply [ALiBi](source), [Gated Linear Units](source), ["The Unpadding Trick"](source), and [FusedLayerNorm](source) and get them to play nicely.
+This folder contains starter code for training various BERT models: **(TBD FILL IN)**. Our goal was to apply a combination of methods from the literature to seriously speed up training time, and to package it in a way that's easy for you to use on your own problems. We apply [ALiBi (Press et al, 2021)](https://arxiv.org/abs/2108.12409v1), [Gated Linear Units (Shazeer, 2020)](https://arxiv.org/abs/2002.05202), ["The Unpadding Trick" (does anyone have a source/name for this? HazyResearch's 2021 MLPerf submission?)](https://github.com/mlcommons/training_results_v1.1/blob/main/NVIDIA/benchmarks/bert/implementations/pytorch/fmha.py), [FusedLayerNorm (NVIDIA)](https://nvidia.github.io/apex/layernorm.html), and [FlashAttention (Tri Dao, 2022)](https://arxiv.org/abs/2205.14135)... and get them to work together. To our knowledge, many of these methods have never been combined before.
 
-Though we achieve a (TBD) speedup over (reference), we have even more tricks up our sleeve which you can only get access to on the [MosaicML Cloud](https://www.mosaicml.com/)!
+Though we achieve a **(TBD FILL IN)** speedup over **(TBD FILL IN)**, there are further speedups which you can only get access to on the [MosaicML Cloud](https://www.mosaicml.com/)!
```

Comment:
Assuming we don't have results yet, so okay to merge in with TBDs.

Reply:
Aren't these called Nested Tensors? https://pytorch.org/tutorials/prototype/nestedtensor.html

Reply:
@jacobfulano PyTorch uniquely calls them that. A more common name I think is "Ragged" Tensors.

`NestedTensor` is a PyTorch object though, and so I don't think we should use that name, because the inputs to our model are not `NestedTensor` s

In [PyTorch MHA](https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html), they do this unpadding trick that we do, but only in the forward pass if you pass in a `NestedTensor` and a lot of other conditions are met.
<img width="963" alt="image" src="https://user-images.githubusercontent.com/7231205/201165463-c97c5cb6-ab2f-4c4b-91ae-56106ba47200.png">